docs/guide/template-filters.md

```diff
@@ -0,0 +1,118 @@
+In this example, a hypothetical page headline will be retrieved from the page data and then piped to the 'replace' filter, along with two arguments. This filter will scan the incoming data and replace any instances of 'foo' with 'bar'. The output of this filter will then pass the result to the upper function, which will return the input as an uppercase string before being output to the page.
+
+>Note: The order of filters can be important, as they are applied sequentially to the input data.
+
+## Apostrophe supplied Filters
+There are several custom filters and filter sets (multiple, sequential filters) used in Apostrophe templating. Rather than have to remember each filter and the correct order, apostrophe exposes them automatically to the templates. In addition, Apostrophe also adds multiple "helper functions" which do not use the `| foo` filter syntax. These are documented with the modules that provide them. Several of the more common of these are within the documentation for the [util module](/reference/modules/util.html#template-helpers).
+
+## Custom template filters
+If you have special template needs you can also construct custom filters for use in Nunjucks templates.
+
+```javascript
+// lib/modules/link-widgets/index.js
+module.exports = {
+  extend: '@apostrophecms/widget-type',
+  options: {
+    label: 'Link to a Page'
+  },
+  fields: {
+    add: {
+      url: {
+        type: 'url',
+        label: 'URL',
+        required: true
+      },
+      label: {
+        type: 'string',
+        label: 'Label',
+        required: true
+      }
+    },
+    group: {
+      links: {
+        label: 'Links',
+        fields: [ 'url', 'label' ]
+      }
+    }
+  },
+  init(self) {
+    self.apos.template.addFilter({
+      stripHttp: self.stripHttp
+    });
+  },
+  methods(self) {
+    return {
+      stripHttp(s) {
+        return s.replace(/^(https?:|)\/\//, '');
+      }
+    };
+  }
+};
+```
+To use this filter you would simply pipe your data to the filter.
+
+```markup
+// lib/modules/link-widgets/views/widget.html
+<section data-link-widget>
+  <h2>{{ data.widget.label }}: {{ data.widget.url | stripHttp }}</h2>
+</section>
+```
+
+## Alphabetical Apostrophe filter reference
+
+### `| build(obj...)`
+
+Given a URL and one or more objects, this filter adds query string parameters for each of the properties in the objects. If objects affect the same parameter, the last object wins. If a parameter is set to `null` or the empty string it is removed from the URL altogether. This is very useful for adding filters to the current URL, respecting other filters already present, without complicated logic.
+
+The `build` filter has additional features which you can read about in the comments of the `index.js` file of the url module.
+
+### `| clonePermanent`
+
+Given JavaScript data, this filter recursively strips out any properties whose names beginning with a `_`, except `_id`. This is used to avoid pushing large joined documents into the DOM.
```

Comment:
"This is used to avoid pushing related documents into the DOM, keeping markup size down." The word `join` is a 2.x-ism.